File: head/motor/zero_detector/zero_detector.go
package zero_detector

import (
	"github.com/cacktopus/theheads/head/motor"
	"github.com/cacktopus/theheads/head/sensor"
	"go.uber.org/zap"
)

type result struct {
	direction motor.Direction
	open      bool
}

type Detector struct {
	sensor sensor.Sensor

	numSteps              int
	directionChangePauses int

	ch     chan result
	logger *zap.Logger
}

func (d *Detector) Name() string {
	return "ZeroDetector"
}

func NewDetector(logger *zap.Logger, sensor sensor.Sensor, numSteps int, directionChangePauses int) *Detector {
	d := &Detector{
		logger:                logger,
		sensor:                sensor,
		numSteps:              numSteps,
		directionChangePauses: directionChangePauses,
		ch:                    make(chan result),
	}

	go d.controller()

	return d
}

func (d *Detector) Act(pos, target int) (motor.Direction, bool) {
	return d.getStep()
}

func (d *Detector) getStep() (motor.Direction, bool) {
	// this may read from the closed channel, which will result in result.open == false
	result := <-d.ch

	return result.direction, !result.open
}

func (d *Detector) step(direction motor.Direction) bool {
	d.ch <- result{
		direction: direction,
		open:      true,
	}

	return false
}

func (d *Detector) pause() (done bool) {
	for i := 0; i < d.directionChangePauses; i++ {
		if done := d.step(motor.NoStep); done {
			return done
		}
	}
	return false
}

func (d *Detector) stepUntil(
	direction motor.Direction,
	targetValue bool,
	stepCount int,
) (done bool) {
	count := 0

	for count < stepCount {
		value, err := d.sensor.Read()
		if err != nil {
			return true // TODO: make this an actual error
		}

		if value == targetValue {
			count++
		} else {
			count = 0
		}

		if done := d.step(direction); done {
			return done
		}
	}

	return false
}

func (d *Detector) controller() {
	defer close(d.ch)

	// 1) find a spot that is off the sensor

	if done := d.pause(); done {
		return
	}

	if done := d.stepUntil(motor.Forward, false, 20); done {
		return
	}

	if done := d.pause(); done {
		return
	}

	if done := d.stepUntil(motor.Backward, false, 10); done {
		return
	}

	if done := d.pause(); done {
		return
	}

	// 2) sweep an entire circle

	var active []int
	for i := 0; i < d.numSteps; i++ {
		value, err := d.sensor.Read()
		if err != nil {
			d.logger.Error("error reading sensor", zap.Error(err))
		}

		if value {
			d.logger.Debug("sensor active", zap.Int("i", i))
			active = append(active, i)
		}
		if done := d.step(motor.Forward); done {
			return
		}
	}

	steps := 0
	if len(active) > 0 {
		d.logger.Debug("active sensor steps", zap.Ints("active", active))
		sum := 0
		for _, pos := range active {
			sum += pos
		}
		steps = sum / len(active)
		d.logger.Info("steps to zero", zap.Int("steps", steps))
	} else {
		d.logger.Error("sensor was not active during scan")
	}

	// Step backward if the path is shorter
	adjustDirection := motor.Forward
	if steps > d.numSteps/2 {
		adjustDirection = motor.Backward
		steps = d.numSteps - steps
	}

	if done := d.pause(); done {
		return
	}

	for i := 0; i < steps; i++ {
		if done := d.step(adjustDirection); done {
			return
		}
	}

	if done := d.pause(); done {
		return
	}
}

func (d *Detector) Finish(controller *motor.Controller) {
	controller.SetCurrentPositionAsZero()
}

File: py/zero_detector.py
import os
from collections import deque

import log
import motors
from head_controllers import Controller, Step
from head_util import DIRECTION_CHANGE_PAUSES

GPIO_PIN = 21
STEPS = 200


class FakeGPIO:
    def __init__(self):
        pass

    def read_int(self) -> int:
        return 1


class GPIO:
    def __init__(self, pin=GPIO_PIN):
        self._pin = pin
        self._gpio_dir = "/sys/class/gpio"
        self._gpio_pin_dir = "/sys/class/gpio/gpio{}".format(self._pin)
        self._export = "/sys/class/gpio/export"
        self._direction = "/sys/class/gpio/gpio{}/direction".format(self._pin)
        self._value = "/sys/class/gpio/gpio{}/value".format(self._pin)

        if not os.path.exists(self._gpio_pin_dir):
            with open(self._export, "w") as f:
                print(GPIO_PIN, file=f)

        with open(self._direction, "w") as f:
            print("in", file=f)

    def read_int(self) -> int:
        with open(self._value) as f:
            v = f.read()

        return int(v)


class ZeroNotFoundError(Exception):
    pass


class ZeroDetector(Controller):
    def __init__(self, gpio: GPIO):
        self.remaining_steps = 1000
        # self._motor = motor
        self._gpio = gpio
        self.gen = self.find_zero()
        self._done = False

    def act(self, *args) -> Step:
        try:
            step = next(self.gen, None)
        except ZeroNotFoundError:
            self._done = True
            return Step.no_step
        if step is None:
            self._done = True
            return Step.no_step
        return step

    def is_done(self) -> bool:
        return self._done

    def finish(self, stepper):
        stepper.set_current_position_as_zero()

    def step_until(self, direction, target_value, count):
        values = deque(maxlen=count)

        while True:
            v = self.read_value()
            values.append(v)
            if len(values) == count and all(v == target_value for v in values):
                return
            yield from self.step(direction)

    def find_zero(self):
        yield from self.step_until(Step.forward, 1, 50)
        yield from self.step_until(Step.no_step, 1, DIRECTION_CHANGE_PAUSES)
        yield from self.step_until(Step.backward, 1, 25)
        yield from self.step_until(Step.no_step, 1, DIRECTION_CHANGE_PAUSES)
        steps_to_zero = yield from self.scan()

        for i in range(steps_to_zero):
            yield from self.step(Step.forward)

        for _ in range(50):
            yield Step.no_step

    def scan(self):
        log.info("scan")
        values = deque(maxlen=STEPS)
        for i in range(STEPS):
            v = self.read_value()
            if v == 0:
                values.append(i)
            yield from self.step(Step.forward)

        if len(values) == 0:
            return 0

        zero_pos = sum(values) / len(values)
        steps = int(zero_pos)
        return steps

    def step(self, direction):
        self.remaining_steps -= 1
        if self.remaining_steps <= 0:
            raise ZeroNotFoundError

        yield direction

    def read_value(self) -> int:
        return self._gpio.read_int()


def main():
    gpio = GPIO(GPIO_PIN)
    motor = motors.setup()
    zd = ZeroDetector(gpio, motor)
    zd.find_zero()
    pass


if __name__ == '__main__':
    main()
